docs/specifications/url.rst

```diff
@@ -0,0 +1,112 @@
+  hostname     = domain | IPv4address | IPv6address
+
+  swiss-number = segment
+
+  version1     = "#v=1"
+
+See https://tools.ietf.org/html/rfc3986#section-3.3 for the definition of ``segment``.
+See https://tools.ietf.org/html/rfc2396#appendix-A for the definition of ``unreserved``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.1 for the definition of ``IPv4address``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.2 for the definition of ``IPv6address``.
+See https://tools.ietf.org/html/rfc1035#section-2.3.1 for the definition of ``domain``.
+
+Versions
+--------
+
+Though all NURLs are syntactically compatible some semantic differences are allowed.
+These differences are separated into distinct versions.
+
+Version 0
+---------
+
+A Foolscap fURL is considered the canonical definition of a version 0 NURL.
+Notably,
+the hash component is defined as the base32-encoded SHA1 hash of the DER form of an x509v3 certificate.
+A version 0 NURL is identified by the absence of the ``v=1`` fragment.
+
+Version 1
+---------
+
+The hash component of a version 1 NURL differs in three ways from the prior version.
+
+1. The hash function used is SHA3-224 instead of SHA1.
+   The security of SHA1 `continues to be eroded`_.
+   Contrariwise SHA3 is currently the most recent addition to the SHA family by NIST.
+   The 224 bit instance is chosen to keep the output short and because it offers greater collision resistance than SHA1 was thought to offer even at its inception
+   (prior to security research showing actual collision resistance is lower).
+2. The hash is computed over the certificate's SPKI instead of the whole certificate.
+   This allows certificate re-generation so long as the public key remains the same.
+   This is useful to allow contact information to be updated or extension of validity period.
+   Use of an SPKI hash has also been `explored by the web community`_ during its flirtation with using it for HTTPS certificate pinning
+   (though this is now largely abandoned).
+3. The hash is encoded using urlsafe-base64 (without padding) instead of base32.
+   This provides a more compact representation and minimizes the usability impacts of switching from a 160 bit hash to a 224 bit hash.
+
+A version 1 NURL is identified by the presence of the ``v=1`` fragment.
+Though the length of the hash string (38 bytes) could also be used to differentiate it from a version 0 NURL,
+there is no guarantee that this will be effective in differentiating it from future versions so this approach should not be used.
+
+It is possible for a client to unilaterally upgrade a version 0 NURL to a version 1 NURL.
+After establishing and authenticating a connection the client will have received a copy of the server's certificate.
+This is sufficient to compute the new hash and rewrite the NURL to upgrade it to version 1.
+This provides stronger authentication assurances for future uses but it is not required.
+
```

Comment:
I think for completeness it would be great to show examples of v1 NURLs (and v0 ones in the previous section).